pkg/ebpf/dns_parser.go

```diff
@@ -0,0 +1,109 @@
+// +build linux_bpf
+
+package ebpf
+
+import (
+	"bytes"
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+)
+
+const maxIPBufferSize = 200
+
+type dnsParser struct {
+	decoder *gopacket.DecodingLayerParser
+	layers  []gopacket.LayerType
+	payload *layers.DNS
+
+	// Cached translation object to reduce allocations
+	cached *translation
+}
+
+func newDNSParser() *dnsParser {
+	payload := &layers.DNS{}
+
+	stack := []gopacket.DecodingLayer{
+		&layers.Ethernet{},
+		&layers.IPv4{},
+		&layers.IPv6{},
+		&layers.UDP{},
+		payload,
+	}
+
+	return &dnsParser{
+		decoder: gopacket.NewDecodingLayerParser(layers.LayerTypeEthernet, stack...),
+		payload: payload,
+		cached:  new(translation),
+	}
+}
+
+func (p *dnsParser) Parse(data []byte) *translation {
+	err := p.decoder.DecodeLayers(data, &p.layers)
+	if err != nil || p.decoder.Truncated {
+		return nil
+	}
+
+	for _, layer := range p.layers {
+		if layer == layers.LayerTypeDNS {
+			return p.parseAnswer(p.payload)
+		}
+	}
+
+	return nil
+}
+
+// source: https://github.com/weaveworks/scope
+func (p *dnsParser) parseAnswer(dns *layers.DNS) *translation {
+	// Only consider responses to singleton, A-record questions
+	if !dns.QR || dns.ResponseCode != 0 || len(dns.Questions) != 1 {
+		return nil
+	}
+	question := dns.Questions[0]
+	if question.Type != layers.DNSTypeA || question.Class != layers.DNSClassIN {
+		return nil
+	}
+
+	var (
+		domainQueried = question.Name
+		records       = append(dns.Answers, dns.Additionals...)
```

Comment:
This could potentially modify the underlying `dns.Answers` array by appending `dns.Additionals` to it if there is space.  I'm not sure how long lived `dns` is and if that could be problematic.

Reply:
Humn good catch. Actually my main concern there is that I might be writing to the shared memory as most of the fields in `*dns.Layer` are byte slices pointing to the socket ring buffer. If (cap > len) I could be writing to the shared memory since append would use the same underlying array. If is not, that means copying the data over and doing a useless allocation. I'll fix that.